Add hack/release.sh script

codenrhoden opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/guide/pull-requests.md#the-pull-request-submit-process and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
4. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
5. If this PR changes image versions, please title this PR "Bump <image name> from x.x.x to y.y.y."
-->

**What this PR does / why we need it**:
Create a new hack/build-images script that can be used for building new
container images for both prod/GA and CI use. This script accepts
arguments to optionally push the images to a registroy, and tags the
images differently if they are designated as only for CI.

This script is meant to be consumed by the top level Makefile in the
future, but also can be used by developers directly.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note

```

codenrhoden commented on hack/build-images.sh:
```diff
@@ -0,0 +1,117 @@
+readonly BASE_REPO=gcr.io/cluster-api-provider-vsphere
+
+# CAPV_MANAGER_IMAGE is the path of the CAPV manager container image
+readonly CAPV_MANAGER_IMAGE=${BASE_REPO}/capv-release/manager
```
Thoughts on `capv-release/manager` for the image name here?

akutz replied:
Hi @codenrhoden,

The buckets are not sub-directories of the base repo. The proposed image would be `gcr.io/cluster-api-provider-vsphere/capv-release/manager`? I was thinking simply `gcr.io/cluster-api-provider-vsphere/manager`.

codenrhoden commented on hack/build-images.sh:
```diff
@@ -0,0 +1,117 @@
+readonly CAPV_MANAGER_IMAGE=${BASE_REPO}/capv-release/manager
+
+# CAPV_MANAGER_IMAGE_CI is the path of the CAPV manager image for CI usage
+readonly CAPV_MANAGER_IMAGE_CI=${BASE_REPO}/capv-ci/manager
```
And on using `capv-ci` as a different bucket for CI images?

akutz replied:
And then for CI images I was thinking `gcr.io/cluster-api-provider-vsphere/ci/xyz`. If you want to keep things in parallel you could use `gcr.io/cluster-api-provider-vsphere/release` for the release images, but I only envisioned this new project as the home for release images at the root of the gcr prefix. I like the long-term thinking though. I guess for release *images* (not other artifacts), I was just thinking that the GCR root itself would be for releases?

codenrhoden commented on hack/build-images.sh:
```diff
@@ -0,0 +1,117 @@
+  local version
+  version=$(git describe --always --dirty --abbrev=8)
+  if [ "${CI_IMAGE}" ]; then
+    IMAGE_NAME=${CAPV_MANAGER_IMAGE_CI}:${version}
```
So this produces image names like `gcr.io/cluster-api-provider-vsphere/capv-ci/manager:v0.3.0-60-g04abe3e1` for CI images, which is slightly different than what is happening today. Just want to point that change out.

akutz commented:
Hi @codenrhoden,

I like the way this is going. A few thoughts:

* For the image paths, why not:
  * `gcr.io/cluster-api-provider-vsphere/release` and
  * `gcr.io/cluster-api-provider-vsphere/ci`
* The `hack` directory has scripts for building machine images, which is why I suggested calling this `hack/release.sh` or something without *image* in the name. That said, I think I'll move the machine image scripts to a sub-dir of `build/images` anyway. 
* I was thinking this script would handle *all* release artifacts -- images, `clusterctl`, etc. Are you thinking it will be specific to the images?

codenrhoden commented:
> * For the image paths, why not:
>   
>   * `gcr.io/cluster-api-provider-vsphere/release` and
>   * `gcr.io/cluster-api-provider-vsphere/ci`

I personally would not like to see things tagged `edge` in a folder called `release`. I would only want to see version-tagged images (e.g. `v0.3.0`, `v0.3.1`, etc.). It almost makes me think more of stable vs nightly "channels", so doing something more akin to:

 * `gcr.io/cluster-api-provider-vsphere/release`
 * `gcr.io/cluster-api-provider-vsphere/edge`
 * `gcr.io/cluster-api-provider-vsphere/ci`

Then, the `release` bucket would have images tagged like `manager:vX.Y.Z` and `latest` pointing the highest numbered release, the `edge` channel would have images tagged like `manager:vX.Y.Z-g{hash}` or whatever with `latest` pointing to the most recent, and `ci` would have the same thing for CI testing and `latest` tag is never applied. Thoughts on that?

> * The `hack` directory has scripts for building machine images, which is why I suggested calling this `hack/release.sh` or something without _image_ in the name. That said, I think I'll move the machine image scripts to a sub-dir of `build/images` anyway.

Yeah, I can definitely rename it to `release.sh`. :)

> * I was thinking this script would handle _all_ release artifacts -- images, `clusterctl`, etc. Are you thinking it will be specific to the images?

No I was thinking of handling everything as well. I was just wondering how we need to handle/package `clusterctl`. See my comment [above](https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/pull/426#pullrequestreview-260246641). Right now, we don't publish `clusterctl` in any way, but we do build an image for it for CI. I don't think we expect people to use `clusterctl` in a container, more to use the binary directly. So we had thought about pushing it to GCS instead. That would be a separate mechanism, but could definitely be in the same script.

codenrhoden commented:
The latest version of the script produces the following sets of images:

```
gcr.io/cluster-api-provider-vsphere/release/manager        latest                                     2e9f259ac604        25 seconds ago      150MB
gcr.io/cluster-api-provider-vsphere/release/manager        v0.3.0                                     2e9f259ac604        25 seconds ago      150MB
gcr.io/cluster-api-provider-vsphere/ci/manager             v0.3.0-60-g2d49c2c5-dirty                  83e4d23c1e1d        4 hours ago         45.7MB
gcr.io/cluster-api-provider-vsphere/edge/manager           latest                                     83e4d23c1e1d        4 hours ago         45.7MB
gcr.io/cluster-api-provider-vsphere/edge/manager           v0.3.0-60-gc825098b-dirty                  83e4d23c1e1d        4 hours ago         45.7MB
```

codenrhoden commented:
I made another update to add a `-l` option to explicitly tag images as "latest". The reason for this is it makes it possible to use the script more flexibly if we have post-submits that are running on a non master branch.  For example, on the master branch, you could have a post-submit that does `hack/release.sh -l -p`, and that will always build the latest and push it. But you could also have a post-submit for **not**-masters that instead runs `hack/release.sh -p` and that would build new images and push them but not tag them as `latest`. This is helpful if we have release/support branches (e.g. a branch named `release-0.3` when we've already release v0.4 and beyond).

codenrhoden commented:
So I think this script is in good shape... Just to help put it into context, here is how I think it would be used, minus all the makefile details...

For pre-submit CI jobs, it should be called as `hack/release.sh -c -p`
For a post-submit job run on every commit to master: `hack/release.sh -l -p`
For a post-submit job that has a branch definition looking for a tag (e.g. `branch: "^v\d+\.\d+\.\d+$"`): `hack/release.sh -p`

Only two  minor things from above to sort out... (1) how to have `latest` tag only moved for `release` channel when the image is not a support/patch release for a lower Major or Minor version
(2) the given regexp for the branch wouldn't catch tags like `v0.4.0-alpha.0`, and I think it should. We would want to standardize that naming convention and capture that too.

/cc @yastij

akutz commented:
Hi @codenrhoden,

This is unfair, but since #412 merged, can you please add support for the `manifests` image as well? Use [`hack/build-container-images.sh`](https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/blob/master/hack/build-container-images.sh) as a reference. Then delete that file :)

codenrhoden commented:
@akutz 

> This is unfair, but since #412 merged, can you please add support for the `manifests` image as well? Use [`hack/build-container-images.sh`](https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/blob/master/hack/build-container-images.sh) as a reference. Then delete that file :)

I definitely can. I lost track of that work, sorry about that.
One question for you, though. It looks like the new manifests image is capable of building manifests for basically all versions of `manager` image. So, under what circumstances should we build/push a new version of the manifests image? It's not something that we are going to change very often, so I hesitate to have it automatically built/pushed alongside of `manager`.

I can make a separate flag for this script to build that image, which would give us the flexibility to do whatever combinations we need, but thinking ahead to Prow configs, I'm wondering when we would trigger a build of it. Only on tags? every commit to master? I don't think new versions are needed for CI, but I don't know... That's the part I'm unclear on.

akutz commented:
> @akutz
> 
> > This is unfair, but since #412 merged, can you please add support for the `manifests` image as well? Use [`hack/build-container-images.sh`](https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/blob/master/hack/build-container-images.sh) as a reference. Then delete that file :)
> 
> I definitely can. I lost track of that work, sorry about that.
> One question for you, though. It looks like the new manifests image is capable of building manifests for basically all versions of `manager` image. So, under what circumstances should we build/push a new version of the manifests image? It's not something that we are going to change very often, so I hesitate to have it automatically built/pushed alongside of `manager`.
> 
> I can make a separate flag for this script to build that image, which would give us the flexibility to do whatever combinations we need, but thinking ahead to Prow configs, I'm wondering when we would trigger a build of it. Only on tags? every commit to master? I don't think new versions are needed for CI, but I don't know... That's the part I'm unclear on.

I think a new `manifests` image should probably be pushed for every `manager` image, even the ones for PRs. The reason being is that eventually (very soon) we want to get rid of the e2e bootstrap spec and generate the manifest for it during the execution of the e2e test. That way we're also testing that the `manifests` images when we test a PR.

Additionally, because of the proposed image structure, we can create a janitor process in Prow that will automatically remove any images at `gcr.io/cluster-api-provider-vsphere/pr/*` that are older than N days.

## `gcr.io/cluster-api-provider-vsphere/pr`

These images are built to test pull requests:

* Image names
  * `gcr.io/cluster-api-provider-vsphere/pr/manager`
  * `gcr.io/cluster-api-provider-vsphere/pr/manifests`
* Image tags
  * `${PROW_JOB_ID}-${BUILD_ID}`
* Test types
  * pre-submit

While these images should be deleted upon completion of the PR, we can also create a janitor process that uses something [like this](https://gist.github.com/ahmetb/7ce6d741bd5baa194a3fac6b1fec8bb7) to delete images older than N days.

## `gcr.io/cluster-api-provider-vsphere/ci`

These images are built from every commit to `master`:

* Image names
  * `gcr.io/cluster-api-provider-vsphere/ci/manager`
  * `gcr.io/cluster-api-provider-vsphere/ci/manifests`
* Image tags
  * `latest`
  * `$(git describe --dirty)`
* Test types
  * post-submit
    * Builds and pushes images
    * Executes current e2e tests against images
  * periodic
    * Executes conformance tests against cluster created with `gcr.io/cluster-api-provider-vsphere/ci/manager:latest`

## `gcr.io/cluster-api-provider-vsphere/release`

These images are built from annotated tags that match a release tag pattern. For example, the following program returns a non-zero exit code if the provided value is not a release tag:

```shell
#!/bin/bash

# Copyright 2019 The Kubernetes Authors.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

set -o errexit
set -o nounset
set -o pipefail

# Change directories to the parent directory of the one in which this
# script is located.
cd "$(dirname "${BASH_SOURCE[0]}")/.."

usage() {
  cat <<EOF
usage: ${0} [TAG]
  Verifies the provided tag is a release tag.

TAG
  If no tag is provided then "git describe --dirty" is used to obtain the tag.

FLAGS
  -h    prints this help screen
  -x    run the examples
EOF
}

while getopts ':hx' opt; do
  case "${opt}" in
  h)
    usage 1>&2; exit 1
    ;;
  x)
    EXAMPLES=1
    ;;
  \?)
    { echo "invalid option: -${OPTARG}"; usage; } 1>&2; exit 1
    ;;
  :)
    echo "option -${OPTARG} requires an argument" 1>&2; exit 1
    ;;
  esac
done
shift $((OPTIND-1))

# The regular expression matches the following strings:
#   * v1.0.0-alpha.0
#   * v1.0.0-beta.0
#   * v1.0.0-rc.0
#   * v1.0.0
# Any occurence of a digit in the above examples may be multiple digits.
REGEX='^[[:space:]]{0,}v[[:digit:]]{1,}\.[[:digit:]]{1,}\.[[:digit:]]{1,}(-(alpha|beta|rc)\.[[:digit:]]{1,}){0,1}[[:space:]]{0,}$'

# Match the tag against the regular expression for a release tag.
match() {
  if [[ ${1} =~ ${REGEX} ]]; then
    echo "yay: ${1}"
  else
    exit_code="${?}"
    echo "nay: ${1}"
    return "${exit_code}"
  fi
}

# Run examples to illustrate valid and invalid values.
examples() {
  local semvers=" \
    v1.0.0-alpha.0 \
    v1.0.0-beta.0 \
    v1.0.0-rc.0 \
    v1.0.0 \
    v10.0.0 \
    v1.10.0 \
    v1.0.10 \
    v10.0.0-alpha.10 \
    v1.10.0-beta.10 \
    v1.0.10-rc.10 \
    1.0.0 \
    v1.0.0+rc.0 \
    v10a.0.0 \
    1.1.0-alpha.1 \
    v1.0.0-alpha.0a"
  set +o errexit
  for v in ${semvers}; do match "${v}"; done
  return 0
}

main() {
  # Get the tag from the remaining arguments or from "git describe --dirty"
  [ "${#}" -eq "0" ] || tag="${1}"
  [ -n "${tag-}" ] || tag="$(git describe --dirty)"

  # Match the tag against the regular expression for a release tag.
  match "${tag}" 1>&2
}

{ [ "${EXAMPLES-}" ] && examples; } || main "${@-}"
```

### Run the examples

```shell
$ hack/match-release-tag.sh -x
yay: v1.0.0-alpha.0
yay: v1.0.0-beta.0
yay: v1.0.0-rc.0
yay: v1.0.0
yay: v10.0.0
yay: v1.10.0
yay: v1.0.10
yay: v10.0.0-alpha.10
yay: v1.10.0-beta.10
yay: v1.0.10-rc.10
nay: 1.0.0
nay: v1.0.0+rc.0
nay: v10a.0.0
nay: 1.1.0-alpha.1
nay: v1.0.0-alpha.0a
```

### Using the command

#### Provide an explicit tag

```shell
$ hack/match-release-tag.sh v1.0.0
yay: v1.0.0
```

#### Omit the tag for `git describe --dirty`

```shell
$ hack/match-release-tag.sh
nay: v0.3.0-77-g8141bf25
```

#### Omit the tag on a release commit

```shell
$ git checkout -b release/v0.3.0 v0.3.0
Switched to a new branch 'release/v0.3.0'

$  hack/match-release-tag.sh
yay: v0.3.0
```

codenrhoden commented:
@akutz I updated the script to build the `manifests` image as well, added a `-k` flag for specifying a path the key file (which would get overridden if `GCR_KEY_FILE` is set), and deleted the old image building script. Go ahead and take another look.

Also adopted your ci/pr path naming.

codenrhoden commented:
/retest

codenrhoden commented:
@akutz One more question for you. We also want this script to push the `clusterctl` binaries to GCS, and we've setup the buckets to do it. Do you think this script should *build* the binary? It seems to me like the Makefile should do it... But if we do that, do we then pass in a path to the binary to this script? Should it assume a location? 

I'm not sure what the right ordering here would be. Ideally the release script could build the binary if it's not already there, and that would work for Prow when we don't have to worry about stale artifacts, but I also want it to work locally as well, so that would mean the script relying on Make to see that a rebuild was necessary? Not sure, looking for your advice here...

Alternatively we could go with what we have here as a first step, get things working with Prow, and then add the binary push after that.

akutz commented:
Hi @codenrhoden,

> One more question for you. We also want this script to push the clusterctl binaries to GCS, and we've setup the buckets to do it. Do you think this script should build the binary? It seems to me like the Makefile should do it... But if we do that, do we then pass in a path to the binary to this script? Should it assume a location?

You can always assume `clusterctl` at `bin/clusterctl` by using either `make clusterctl` or `make clusterctl-in-docker`. The latter will produce a binary that can be produced locally with the same checksum, but either is fine.

> I'm not sure what the right ordering here would be. Ideally the release script could build the binary if it's not already there, and that would work for Prow when we don't have to worry about stale artifacts, but I also want it to work locally as well, so that would mean the script relying on Make to see that a rebuild was necessary? Not sure, looking for your advice here...

Since `clusterctl` pulls in the same types that go into the vSphere manager, it's fine to use the same `run_if_changed` value. For example:

```yaml
run_if_changed: '^(cmd|config|pkg|vendor)/)|Dockerfile'
```

akutz commented:
Hi @codenrhoden,

Did you have any additional questions about the above response?

codenrhoden commented:
@akutz 

Nothing from that response, but I thought of one more thing I'd like us to agree on. As you pointed out, right now the docs say to download `clusterctl` from the GitHub release page. That page has different versions, and offers both a `darwin_amd64` and `linux_amd64` binary.

Do we want Prow to push both a Linux and MacOS binary? It would seem useful... But if so, we need to modify the Makefile to build both and place them in `bin` so we can send them to GCS. Introducing this script is going to necessitate changes to the Makefile anyways, so that's no big deal... Just wanted to hear your thoughts on it. I think I will have the script look for the file(s) (as defined below) and send whatever it finds that matches. If there is at least one match, it's successful. If no builds are found, it will exit with error. Thoughts?  Might make more sense after I implement it.

Second question is how to attach the version. Right now we have these three buckets set up:

`capv-ci`
`capv-pr`
`capv-release`

I don't think we want to make subdirectories within those, do you? So what do you see the filenames being?

I'm thinking... `clusterctl-{VERSION}.linux_amd64`. 

What do you think?

akutz commented:
Hi @codenrhoden,

> Nothing from that response, but I thought of one more thing I'd like us to agree on. As you pointed out, right now the docs say to download clusterctl from the GitHub release page. That page has different versions, and offers both a darwin_amd64 and linux_amd64 binary.

If we build it with CI we would update the docs -- or not. We can just upload the `clusterctl` binary to GitHub after a release.

> darwin

It wouldn't be a big deal to add that to the script. I have this in my `.scratch` directory to do just that. It's how those binaries on GitHub got there.

```shell
for goos in darwin linux; do GOOS="${goos}" make clusterctl-in-docker; done
```

> `clusterctl-{VERSION}.linux_amd64`

Hmm, a few things:

* I don't the version should ever need to be part of the file name as it should be part of the path structure since I would think we'd want to avoid completely flat release directories
* To that note, what about omitting the OS and platform from the file name as well and using a structure similar to the one used by the Kubernetes buckets?

    ```shell
    gs://capv-release/v0.4.0/bin/linux/amd64/clusterctl
    gs://capv-release/v0.4.0/bin/darwin/amd64/clusterctl
    ```

I'm not married to either idea, but I do like of following k/k structure when possible. Thoughts?

codenrhoden commented:
> ```shell
> for goos in darwin linux; do GOOS="${goos}" make clusterctl-in-docker; done
> ```
> 
Ha, this is exactly what I already did. =)

> > `clusterctl-{VERSION}.linux_amd64`
> 
> Hmm, a few things:
> 
> * I don't the version should ever need to be part of the file name as it should be part of the path structure since I would think we'd want to avoid completely flat release directories
> * To that note, what about omitting the OS and platform from the file name as well and using a structure similar to the one used by the Kubernetes buckets?
>   ```shell
>   gs://capv-release/v0.4.0/bin/linux/amd64/clusterctl
>   gs://capv-release/v0.4.0/bin/darwin/amd64/clusterctl
>   ```
> 
> I'm not married to either idea, but I do like of following k/k structure when possible. Thoughts?

Hmm... yeah I could go either way. i guess in the absence of strong opinion, it makes sense to follow k/k.

codenrhoden commented:
@akutz This is ready for a thorough look. I've tested that it works locally by building and pushing "PR" images. In all other scenarios I commented out the push portion to as not to actually push anything to CI or Release areas, but it all looked good.

I think the only thing I'm not sure of is if, when we are running from Prow, we will need to do anything special to login for the `gsutil` commands to work. The GCR stuff is all set, but I'm not sure about the `gsutil` command.

akutz commented:
Hi @codenrhoden,

I *believe* commands like `gsutil cp` and `gsutil rsync` are subject to the same `gcloud auth login` used ahead of pushing container images.

It would be easy enough to test this in a clean environment by running it inside a container.

akutz commented on hack/release.sh:
```diff
@@ -0,0 +1,219 @@
+  -l    tag the images as "latest" in addition to their version
+        when used with -p, both tags will be pushed
+  -p    push the images to the public container registry
+  -t    the images are solely for the use by CI/testing
```
Hi @codenrhoden,

I'm wondering if this should stand for `type` of build, and have its valid values be `ci|release|pr`. That would enable `case/esac` for the build type:

```shell
case "${BUILD_RELEASE_TYPE}" in
  ci)
    ;;
  release)
    ;;
  pr)
    ;;
esac
```

So the changes might look like:

```shell
BUILD_RELEASE_TYPE ="${BUILD_RELEASE_TYPE-}"

# If BUILD_RELEASE_TYPE is not set then check to see if this is a PR
# release build. This may still be overridden below with the "-t" flag.
if [ -z "${BUILD_RELEASE_TYPE}" ]; then
  if hack/match-release-tag.sh >/dev/null 2>&1; then
    BUILD_RELEASE_TYPE=release
  else
    BUILD_RELEASE_TYPE=pr
  fi
fi

...

function usage() {
  cat <<EOF
usage: ${0} [FLAGS]
  Builds and optionally pushes new images for Cluster API Provider vSphere (CAPV)
FLAGS
  -h    show this help and exit
  -k    path to GCR key file. Overriden if GCR_KEY_FILE env var is set
  -l    tag the images as "latest" in addition to their version
        when used with -p, both tags will be pushed
  -p    push the images to the public container registry
  -t    the build/release type (defaults to ${BUILD_RELEASE_TYPE})
EOF
}

...

# Start of main script
while getopts ":hk:lpt:" opt; do
  case ${opt} in
    h)
      usage 1>&2; exit 1
      ;;
    k)
      GCR_KEY_FILE="${OPTARG}"
      ;;
    l)
      LATEST=1
      ;;
    p)
      PUSH=1
      ;;
    t)
      BUILD_RELEASE_TYPE="${OPTARG}"
      ;;
    \?)
      { echo "invalid option: -${OPTARG}"; usage; } 1>&2; exit 1
      ;;
    :)
      echo "option -${OPTARG} requires an argument" 1>&2; exit 1
      ;;
  esac
done
shift $((OPTIND-1))

# Verify the GCR_KEY_FILE exists
[ -e "${GCR_KEY_FILE}" ] || fatal "key file ${GCR_KEY_FILE} does not exist"

# Validate build/release type.
case "${BUILD_RELEASE_TYPE}" in
  ci|pr|release)
    # do nothing
    ;;
  *)
    fatal "invalid BUILD_RELEASE_TYPE: ${BUILD_RELEASE_TYPE}"
    ;;
esac
```

akutz commented on hack/release.sh:
```diff
@@ -0,0 +1,219 @@
+  Builds and optionally pushes new images for Cluster API Provider vSphere (CAPV)
+
+FLAGS
+  -h    Show this help and exit
```
`/Show/show` to be case-consistent with other flags

akutz commented on hack/release.sh:
```diff
@@ -0,0 +1,219 @@
+
+FLAGS
+  -h    Show this help and exit
+  -k    Path to GCR key file. Overriden if GCR_KEY_FILE env var is set
```
`/Path/path` to be case-consistent with other flags

akutz commented on hack/release.sh:
```diff
@@ -0,0 +1,219 @@
+}
+
+function build_images() {
+  if [ "${RELEASE}" ]; then
```
Also, see my above comment about build/release type. This may end up being a different change altogether if you like that idea.

akutz commented on hack/release.sh:
```diff
@@ -0,0 +1,219 @@
+}
+
+function push_images() {
+  [ "${MANAGER_IMAGE_NAME}" ] || exit 1
```
Hi @codenrhoden,

This should be:

```
[ "${MANAGER_IMAGE_NAME}" ] || { echo "MANAGER_IMAGE_NAME not set" 1>&2; exit 1; }
```

akutz commented on hack/release.sh:
```diff
@@ -0,0 +1,219 @@
+
+function push_images() {
+  [ "${MANAGER_IMAGE_NAME}" ] || exit 1
+  [ "${MANIFESTS_IMAGE_NAME}" ] || exit 1
```
Hi @codenrhoden,

This should be:

```
[ "${MANIFESTS_IMAGE_NAME}" ] || { echo "MANIFESTS_IMAGE_NAME not set" 1>&2; exit 1; }
```

akutz commented on hack/release.sh:
```diff
@@ -0,0 +1,219 @@
+      continue
+    fi
+    echo "copying clusterctl version ${VERSION} for ${os} to ${bucket}"
+    gsutil cp bin/clusterctl."${os}"_amd64 gs://"${bucket}"/"${VERSION}"/bin/"${os}"/amd64/clusterctl
```
Hi @codenrhoden,

Do we want to add a shasum?

codenrhoden replied:
Yeah, will a SHA256 sum.

akutz commented:
Removed [WiP] and adding

/hold

that can be removed once the PR has been reviewed.

codenrhoden commented on hack/release.sh:
```diff
@@ -0,0 +1,219 @@
+  fi
+
+  # Manager image
+  echo "building" "${MANAGER_IMAGE_NAME}":"${VERSION}"
```
Nope, just something I did wrong. =) will fix.

codenrhoden commented on hack/release.sh:
```diff
@@ -0,0 +1,219 @@
+
+FLAGS
+  -h    Show this help and exit
+  -k    Path to GCR key file. Overriden if GCR_KEY_FILE env var is set
```
Isn't the same as saying that the `GCR_KEY_FILE` env var overrides it?

akutz replied:
Hi @codenrhoden,

Yes, but one prints the value of the environment variable as a reminder to people what file will be used.

codenrhoden replied:
ah, now I see your point. Happy to add.

codenrhoden commented on hack/release.sh:
```diff
@@ -0,0 +1,219 @@
+
+# Get the version from the repo
+VERSION=$(git describe --dirty --always 2>/dev/null)
+hack/match-release-tag.sh 2>/dev/null && RELEASE=1
```
It hasn't been, no. I've been testing this against dirty repos, tagged and untagged commits, and it's produced the right result each time.

codenrhoden commented:
Okay, I did a bunch of the cleanups and add the SHA sum.  I have not addressed anything to do with `gcloud auth` or the KEYFILE, will have to do that next/later.

As for the build type, I may take the approach to iterate on that later. =) But we can discuss more after the weekend.

codenrhoden commented:
Hi @akutz,

I incorporated your feedback and added the last piece of this - the `gcloud auth` integration. If someone specifies a keyfile, we use that keyfile to activate a service account and then revoke that account when we are done. We *only* do this if we are pushing images. If no key file is specified, we assume the user has already logged in (meant for local devs). 

I'm not sure if when we are running this on Prow we need to explicitly install the gcloud auth docker helper or not.

codenrhoden commented:
> I'm not sure if when we are running this on Prow we need to explicitly install the gcloud auth docker helper or not.

I'm guessing we don't need to do anything here, otherwise the existing `docker push`'s wouldn't work.

akutz commented:
/lgtm
/approve

This is great @codenrhoden, thank you!

akutz commented:
/hold cancel
